### DISCUSSION TITLE
关于hy搞负载均衡故障转移的一些探讨

### DISCUSSION THREAD

**Discussion by Tmalll** [apernet/hysteria#420] [Ideas]
这两天我用openwrt上的passwall配合本地hysteria-client搞了个hy的负载均衡+故障转移.
拓扑如下 [PC]---透明代理---[openwrt_passwall]---负载均衡--[本地hysteria-1/2/3/4]---hy通道--[VPS上的hysteria服务器]

passwall一下简称pw
hysteria简称hy

发现几个难点:
1. pw的负载均衡用的haproxy, 貌似只能用tcp负载均衡, udp的行?
2. 基于1导致需要先在本地用hy客户端把连接转成s5
3. pw里haproxy识别[掉线]和[在线]时只能识别直连, 被转发或者后端服务器宕机都无法识别出来,从而进行故障转移切换.

本地debian上搞了4个hy客户端分别对应50100/50200/50300/50400四个s5端口
![image](https://user-images.githubusercontent.com/40192806/187432939-c0c92e6a-2ee5-494c-abde-39f919979f73.png)

这是pw里负载均衡设置的逻辑
![image](https://user-images.githubusercontent.com/40192806/187433233-1e0644f5-45e7-4756-93a9-a5a8fbf33e7e.png)

基于上面问题3, 如果1.29这台服务器出问题故障能转移,501~504个端口出问题也可以故障转移, 但是如果后端vps出问题就没法自动切换节点了, 目前我想到一个办法, 在1.29上写个ping的脚本,自动ping后端vps,如果不通则关闭对应的hy守护服务,从而达到关闭端口,ping的通在自动打开hy服务, 然后放到计划任务里每几分钟运行一次.

如果hy能按下图出个功能开关, 连不上主服务器或者故障的时候,自动关闭下面的s5/转发端口就好了.
![image](https://user-images.githubusercontent.com/40192806/187431007-c9748e53-c03b-4d48-8ee4-a8c2ca6a040e.png)

各位大佬还有啥更好的实现办法吗?

**Comment by haruue** [comment]
iptables 的 INPUT 链适用于本地 loopback 通信， 
加一条 iptables 应当可以起到和关闭端口一样的作用
```
iptables -I INPUT -p tcp --dport 1080 -j REJECT --reject-with tcp-reset
```

**Comment by haruue** [comment]
考虑了下其实没那么麻烦。。 加个选项让 hysteria 在 quic 断开的时候直接退出就行了。。
在这个位置。
https://github.com/HyNetwork/hysteria/blob/master/pkg/core/client.go#L171

然后外面用脚本（或者 systemd）自动重启 hysteria， 而 hysteria 的话在成功连接上服务器之前是不会监听 socks5 端口的。

**Reply by Tmalll** [reply]
这个办法也不错, 昨天我也搞了下systemd, 发现它不会自动退出就放弃了... 希望下个版本能加上这个功能, 做成一个开关也行.

**Comment by hujiao80** [comment]
现在用的是直接改PW的app.sh中haproxy healthcheck的配置，模拟socks5的连接握手过程来做线路健康检查
```
 option tcp-check
 tcp-check connect
 tcp-check send-binary 05020002
 tcp-check expect binary 0500
 tcp-check send-binary 050100030d7777772e62616964752e636f6d01bb #用www.google.com做检测网址
 tcp-check expect binary 05000001
```

**Reply by Tmalll** [reply]
> 现在用的是直接改PW的app.sh中haproxy healthcheck的配置，模拟socks5的连接握手过程来做线路健康检查
> 
> ```
>  option tcp-check
>  tcp-check connect
>  tcp-check send-binary 05020002
>  tcp-check expect binary 0500
>  tcp-check send-binary 050100030d7777772e62616964752e636f6d01bb #用www.google.com做检测网址
>  tcp-check expect binary 05000001
> ```

老哥这个文件在哪里? 还是pw的配置里面?

**Reply by hujiao80** [reply]
/usr/share/passwall
